### ISSUE TITLE
Can't install TAC-LS without picking the Stock-Config

### ISSUE THREAD

**Issue by Dazpoet** [KSP-CKAN/CKAN-GUI#95]
Bad title aside...

So as we all know (who am I kidding) the GUI always seems to autopick the first option when a dependency includes multiple choices. This seems to be due to https://github.com/KSP-CKAN/CKAN-core/pull/82

TACLS depends on `TACLS-Config` which is provided by `TACLS-Config-RealismOverhaul` and `TACLS-Config-Stock` The former is the first choice (assuming due to alpabetical ordering?) and in turn depends on `RealismOverhaul` which isn't updated to KSP 1.0.2 yet.

Since the GUI autopicks the first choice the effect is that rather than present the user the ability to choose the Stock-Config when attempting to install TACLS it just won't let the user do it at all. Choosing to instead install the Stock-Config will generate an installation of both it and TACLS.

Problem does not persist in Commandline client where the user is presented with a choice.

Reproduction steps:

Requirement: An installation where TAC-LS is installable but RealismOverhaul is not, at time of writing this is true for 1.0.2 of KSP

#### GUI

1) Check TAC-LS in the GUI - Get no Changeset tab
2) Check TAC-LS Stock Config - Get a Changeset tab which includes TAC-LS

#### Commandline
1) `ckan install TACLS`
2) Be presented with a choice where RO leads to a failure (as it should) and Stock successfully installs

Thoughts:

Could this GUI behaviour be related to the same code which presents the issues described in https://github.com/KSP-CKAN/CKAN-support/issues/103 and https://github.com/KSP-CKAN/CKAN-support/issues/104 or are those unrelated?

**Comment by Dazpoet** [comment]
Another example of something I think might be related.

When installing Astronomers Pack the `AstronomersPack` identifier has the following dependency
```
    "depends": [
        {
            "name": "AstronomersPack-Clouds"
        }
```
In commandline the installation of AstronomersPack correctly lets me choose between High/Low/Medium quality clouds.

In GUI it preselects High quality and the only way around this is to manually select the cloud quality by checking its box prior to installation.

**Comment by RNeilsen** [comment]
Having the exact same issue (CKAN v1.6.16, KSP 1.0.2). My mod list in CKAN currently includes

* TAC Life Support (v0.11.1.20)
* TAC Life Support (TACLS) - stock config (v0.11.1.20)

If I tick TAC Life Support first, I get the error message. If I tick the 'stock config' one first, and *then* tick TAC Life Support (**without** installing in between), it works perfectly and I can install both mods at the same time.

**Comment by pjf** [comment]
@Dazpoet : Do you know if the GUI always used to pick the "first option" when dependencies came into play? This really feels like it's a regression; the whole point of depends/provides relationships so the user gets to choose an option (or the relationship resolver does if there's only one non-conflicting choice), rather then the user getting an arbitrary option that probably doesn't suit their needs.

**Comment by RichardLake** [comment]
@pjf I've taken a look. AlexanderDzhoganov did write code to allow the user to choose provides. Haven't gone through the entire history but it seems to have never been reachable. It is called in the install phase if a toomanyprovides error is throw, however the changeset generation is called with the without_toomany_provides option which has it choose the first provide.

**Comment by Dazpoet** [comment]
@pjf I'm afraid I cannot recall if it was always like this or not. @RichardLake the only PR which I know of offhand that dealt with choices and autopicking is https://github.com/KSP-CKAN/CKAN-core/pull/82 and I'm far from sure that it's related.

My metapackage from https://github.com/KSP-CKAN/CKAN/issues/913#issuecomment-104149980 still seems to run straight through CKAN-GUI without doing anything so that might be used as a testcase if it's the same issue (which I'm not sure it is).

**Comment by Dazpoet** [comment]
@postremus reported [here](https://github.com/KSP-CKAN/CKAN/issues/953#issuecomment-105061432) the this behaviour did not exist in CKAN 1.5.6 so I did some further testing and up to 1.6.2 TACLS installs fine and presents a choice in GUI even on KSP 1.0.2 installs.

CKAN 1.6.3 doesn't even start on my testinstalls and iirc it was a borked version which dealt with autoupdates as was 1.6.4. First version I got running after 1.6.2 was CKAN 1.6.5 in which the issue detailed in this report also appears.

To sum it up;

* Issue does not appear in CKAN 1.5.6 through 1.6.2
* CKAN 1.6.3 does not start on my testinstall
* CKAN 1.6.4 is not listed under releases and has not been tested
* Issue first observerd in CKAN 1.6.5

**Comment by Postremus** [comment]
tacls installs fine with ckan 1.6.3, using a ksp 1.0.2 install.

**Comment by pjf** [comment]
Excellent! For the GUI:

- v1.6.5 was 42aa1a7ed84cca3f4d28a7b095ee13343e24402f
- v1.6.3 was 4d65585215814fc4ceee223b4973d10da6add28c
- v1.6.2 was c2b6adff74cc116c1efb8b2f0db01c1772b4bf8e

However I'm having a hard time finding anything that might cause this [given the commits in this timeframe](/KSP-CKAN/CKAN-GUI/compare/c2b6adff74cc116c1efb8b2f0db01c1772b4bf8e...42aa1a7ed84cca3f4d28a7b095ee13343e24402f), so I'm wondering if it was a core change that affected the GUI somehow...

**Comment by pjf** [comment]
In the core:

* v1.6.5 was b1a2e82ee269d21b945dcb29856c123742f3f352
* v1.6.3 was 0ec39ec9b1d45e3b5e93623deed4dc5ba1eb0c38

The most obvious commit would be https://github.com/KSP-CKAN/CKAN-core/commit/72577c59d88e29ec1f326cbed58b35921fbb5676 , which is clearly intended to be used in headless mode, but also affects the GUI since it runs without our dear friend the TooManyProvides kraken.

I'm not exactly sure this is the best solution for `--headless`, if a mod is being installed and a choice needs to be made, then to me it seems reasonable to bail out rather than to just pick one. If we *did* want auto-picking we'd likely want something similar to Debian's priority field, so we can avoid issues like what we've seen here. (At least, I believe that's how Debian solves the issue; I've not properly checked myself.)

It looks like this work was done as part of KSP-CKAN/CKAN-core#82 as @Dazpoet suggests.

I'm not sure of the context in which this was made; I'm guessing it might be for automated testing with jenkins, but that's a guess at best. I'm certainly happy to revert it with we can confirm that doing so restores the intended user behaviour.

Summoning @AlexanderDzhoganov and @aw1621107 in case they can shed further light. Obviously we'd love to fix this in the GUI while preserving whatever

**Comment by aw1621107** [comment]
*sigh* I was afraid that I messed something up... Sorry about that.

@pjf you're right, this was for automated testing. Before this change was made, CKAN in --headless mode would refuse to pick a config if there was more than one option, which would result in a broken install and a failed build. The change was intended to ensure that mods with more than one config option were installed properly, so if a build failed it would be for a more useful reason. Because we just needed to see if the install worked, the particular config chosen didn't really matter (or so I thought). So I just had it default to the first choice.

I was told somewhere that --headless was only used for automated builds, so to see this breaking something in the GUI is surprising.

**Comment by pjf** [comment]
> sigh I was afraid that I messed something up... Sorry about that.

Psh! No apologizing here! Seriously. If bugs get into the code, we should be asking how we can improve our processes.

> I was told somewhere that --headless was only used for automated builds, so to see this breaking something in the GUI is surprising.

The gotcha is that both headless mode and the GUI run with `without_toomanyprovides_kraken` set, probably so it can get back partial solutions and then quiz the user what to do from there. (At some point I'd *love* to rename the variable to something like `return_partial_on_failure`, which I think is *much* more descriptive of what it's doing.)

I think it would be sensible to take a good look at our relationship resolver options, to make sure they're all necessary and appropriately descriptive. We could add a `force_...` option to force the picking of at least one option (which headless mode can invoke). Ideally, if we force-picked a mod, we'd use the suggests/recommends lists from the mod itself to break ties.

Alternatively, we could add some sort of mark-up to our metadata so that our test harnesses can install additional mods before the one we're testing, allowing some dependencies to be pre-resolved. (This means users in headless mode will still need to specify which alternative they want, which isn't a bad thing at all.)

Almost 5am for me. More when I'm awake. :)

**Comment by Dazpoet** [comment]
"Alternatively, we could add some sort of mark-up to our metadata so that our test harnesses can install additional mods before the one we're testing, allowing some dependencies to be pre-resolved" -pjf

This would pretty much be the awesomest metadata tool in existance and would solve situations like https://github.com/KSP-CKAN/NetKAN/pull/1145 and https://github.com/KSP-CKAN/NetKAN/pull/1264 where the lexical installation order of mods breaks builds which are not faulty due to dependencies not being installed in the correct order.

**Comment by aw1621107** [comment]
Hmmm... Would options be out of the question? Sounds like the first one is the faster/easier fix, and the second allows for some more flexibility and options outside of this issue.

**Comment by mgsdk** [comment]
> Alternatively, we could add some sort of mark-up to our metadata so that our test harnesses can install additional mods before the one we're testing, allowing some dependencies to be pre-resolved. -pjf

I've been thinking about changing the build scripts, at least for NETkan, such that for PR's with more than one netkan/ckan file:
* The current metadata is downloaded to storage and extracted.
* All but one of the ckan files are copied into the folder, and the metadata repackaged.
* The default repository is exchanged with the local metadata in `ckan.exe update`.

This would at least solve the two issues @Dazpoet is referencing, since all the other packages would be available in the metadata. But I'm not sure if what @pjf and @aw1621107 are talking about with regards to testing is the same as this?

**Comment by pjf** [comment]
> Hmmm... Would options be out of the question? Sounds like the first one is the faster/easier fix, and the second allows for some more flexibility and options outside of this issue.

Can you explain to me how you'd envision options?

> This would pretty much be the awesomest metadata tool in existance and would solve situations like KSP-CKAN/NetKAN#1145 and KSP-CKAN/NetKAN#1264 where the lexical installation order of mods breaks builds which are not faulty due to dependencies not being installed in the correct order.

My brain is made of meat and does a poor job when faced with complexity. We should never require that mods be installed in *any* particular order, provided everything is done as part of the same transaction. What are we encountering that's requiring ordering? My idea of priority is that that it can act as a tie-breaker when we have to choose between mods without a human present, but not that it will affect our installation order in any way.

Having said that, the current behaviour of jenkins in trying to install *all* the mods in a PR to the same virtual KSP is definitely broken, because PRs often come with conflicting changes that can't be installed alongside each other. @mgsdk, if you're suggesting we change things so we wipe/renew the virtual KSP for each file, then that would be fantastic. But I'm not sure I follow the steps you've listed, so I fear I may be confused.

(On that note, I'm very much looking forward to moving jenkins over to our AWS infrastructure, because we can then keep a local  copy of all the mods and never have to wait for downloads again.)

Note: I've only had a few hours sleep last night, so my comprehension abilities are almost certainly poor.

**Comment by aw1621107** [comment]
> Hmmm... Would options be out of the question? Sounds like the first one is the faster/easier fix, and the second allows for some more flexibility and options outside of this issue.

Oops. Should be "Would doing both options..."

**Comment by mgsdk** [comment]
> My brain is made of meat and does a poor job when faced with complexity. We should never require that mods be installed in any particular order, provided everything is done as part of the same transaction. What are we encountering that's requiring ordering? My idea of priority is that that it can act as a tie-breaker when we have to choose between mods without a human present, but not that it will affect our installation order in any way.

When we do changes like splitting a package up into a core package with the dll and a config package, Jenkins will fail the tests *if* the config is tested prior to the core package, as it is not available in the repository. I wanted to circumvent this by injecting all the other mods in the PR but the one being tested into the registry, such that the dependency resolver should be able to handle it. However, it makes sense to wipe the KSP dir as you said between each test as well. Another thing worth expanding on would be to actually match the min KSP version in the ckan file to test, instead of always going with the latest.

**Comment by pjf** [comment]
On phone, but expressing strong agreement.

Just to check, is there a reason why we can't inject all packages in a PR, rather than "all but the one we're testing"? That strikes me that it should have the same effect, but with less complexity.

**Comment by mgsdk** [comment]
> Just to check, is there a reason why we can't inject all packages in a PR, rather than "all but the one we're testing"? That strikes me that it should have the same effect, but with less complexity.

I believe we circumvent the relationship resolver to some degree by invoking the install command with the `-c` argument (i.e. install from this .ckan file). I guess we could get the same functionality by invoking the install command with `ckan.exe install mod=version` and injecting all the metadata at the beginning.

I have an attempt at implementing the changes in https://github.com/KSP-CKAN/NetKAN/pull/1423.

**Comment by pjf** [comment]
Just a quick note that I'm going to be reverting the relevant commit and pushing a new release. Bugs that our developers have to endure are better than bugs our users have to endure.

I'm hoping to go over @mgsdk 's changes in the build script as well. My brain isn't running at full speed, but changing the code jenkins uses for building is pretty darn low-risk, as there's no user visibility if things go wrong.

**Comment by pjf** [comment]
Richard has https://github.com/KSP-CKAN/CKAN-core/pull/139 , but on my system it causes the GUI to lock rather than crash. So I'm still progressing towards reverting the relevant commit for the time being.
